### PR TITLE
Add `depends_on([c,cxx,fortran])`

### PR DESCRIPTION
**Pull request by haampie**
generated with https://github.com/haampie/spack-language-to-package-py: based on file extensions in archives w/o looking at file contents or build systems in detail.

the idea is people remove `# generated` when it's verified.

### CODE REVIEW

**Comment by alalazo** [comment]
Removed from being a draft to check concretization in pipelines

**Comment by tgamblin** [comment]
@alalazo can you time concretization on this?

**Comment by tgamblin** [comment]
Based on [this](https://github.com/haampie/spack-language-to-package-py/blob/master/update_languages.py#L18-L20):

```python
C_EXT = {".c"}
CXX_EXT = {".cpp", ".cc", ".cxx", ".c++"}
FORTRAN_EXT = {".f", ".f77", ".f90", ".f95", ".f03", ".f08"}
```

I think we're missing a few suffixes -- these are what `cloc` uses:

```console
> cloc --show-lang=c
C                          (c, cats, ec, idc, pgc)
> cloc --show-lang=c++
C++                        (C, c++, cc, CPP, cpp, cxx, h++, inl, ipp, pcc, tcc, tpp)
> cloc --show-lang="fortran 77"
Fortran 77                 (F, F77, f77, FOR, FTN, ftn, pfo)
> cloc --show-lang="fortran 90"
Fortran 90                 (F90, f90)
> cloc --show-lang="fortran 95"
Fortran 95                 (F95, f95)
```

And I guess I would also add `hpp` for C++ as I see that frequently.  Once @alecbcs gets the source mirror fully downloaded here, can we go and re-run your script and do some sampling to check it?

**Comment by haampie** [comment]
>  Once @alecbcs gets the source mirror fully downloaded here, can we go and re-run your script and do some sampling to check it?

Don't see the point of waiting for that. The `# generated` comment is enough to update packages at a later stage, I doubt it can be improved.

I've force pushed the PR with the patch generated by https://github.com/haampie/spack-language-to-package-py/actions/runs/9952936512/job/27495303284, the number of [changed files is 89](https://github.com/spack/spack/compare/9d19d3a314c8655f4aecb6e1c26a44f98717ec2e..88eab7fe18cb950bae2a8edac53a8e21b2d0889e), half of which is about .C classified as C++ now instead of C, so that leaves about 1.5% of package changes due to flaky downloads compared to last run. That's likely already below the false positive threshold, we don't need 100% accuracy on 90% reliable data...

IMO let's merge this now, it's easy enough to fix things later. I don't want to rebase this.

File: var/spack/repos/builtin/packages/adios2/package.py [hunk 1]
```diff
@@ -45,6 +45,10 @@ class Adios2(CMakePackage, CudaPackage, ROCmPackage):
     version("2.4.0", sha256="50ecea04b1e41c88835b4b3fd4e7bf0a0a2a3129855c9cc4ba6cf6a1575106e2")
     version("2.3.1", sha256="3bf81ccc20a7f2715935349336a76ba4c8402355e1dc3848fcd6f4c3c5931893")
 
+    depends_on("c", type="build")  # generated
+    depends_on("cxx", type="build")  # generated
+    depends_on("fortran", type="build")  # generated
```
Comment (alecbcs) [inline var/spack/repos/builtin/packages/adios2/package.py:50]: We might want to update the generation script to look for `variant("<lang>")` in the packages. `adios2` only depends on `fortran` when building the fortran bindings with `+fortran`.

**Comment by tgamblin** [comment]
> I doubt it can be improved.

Ok, I agree with this *with the additional suffixes*. Thanks!

File: var/spack/repos/builtin/packages/adios2/package.py (hunk 1 above)
Comment (tgamblin) [reply var/spack/repos/builtin/packages/adios2/package.py:50]: I'm ok with doing this in a follow-on PR to get the initial state in, provided CI is passing.  I do think this is a good idea.

**Review by tgamblin — DISMISSED** [review]

**Comment by haampie** [comment]
fwiw e9decdd was cherry-picked and b539eb5aabe2b377395dc276761e372f6ed3a774 on develop now, it still shows here as a separate commit

**Comment by tgamblin** [comment]
Trying on a few of the more complicated packages (`hdf5`, `mfem`, `lbann`) I don't see much if any difference in the timings for this PR and `develop`. 🎉

**Comment by tgamblin** [comment]
It looks like SuperLU-dist is failing with some Fortran-related error.

**Comment by alalazo** [comment]
I confirm that I see no significant differences in concretization times:

* **Spack:** 0.23.0.dev0 (89c0b4accf0655d280f6c84a332cc95a03e90aef)
* **Python:** 3.11.5
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

[radiuss.pr.csv](https://github.com/user-attachments/files/16261972/radiuss.pr.csv)
[radiuss.develop.csv](https://github.com/user-attachments/files/16261973/radiuss.develop.csv)
[radiuss.txt](https://github.com/user-attachments/files/16261974/radiuss.txt)

radiuss

A few specs seems slower, a few faster. Notably, `py-deephyper` (which is kind of unstable in concretization time) is 30 sec. faster in this branch using gcc. Maybe adding information on fortran/cxx virtuals helped concretizaton.

I would be for merging this asap,

**Comment by haampie** [comment]
apart from the significant differences no significant differences?

**Comment by alalazo** [comment]
> apart from the significant differences no significant differences?

I mean, there are no trends for things getting generally slower or faster due to this PR :wink:

**Comment by haampie** [comment]
I see tons of rebuilds. Non-determinism?

Edit: nope it was 40b390903dae1d263dc6106926b5aac5266d5201

**Comment by haampie** [comment]
FWIW, there is non-determinism, but not due to this PR :)

These two specs

```
warpx ^openpmd-api@0.14
warpx ^openpmd-api@0.15
```

have the same "version badness", because

- `openpmd-api@0.14` is 3 versions from latest but does not depend on toml11
- `openpmd-api@0.15` is latest, but depends `toml11` 3 versions from latest

:tada: 

not a blocker here, just happened to be that `warpx ^openpmd-api@0.14` was picked and that exposed a missing compat bound.

**Comment by haampie** [comment]
`cairo` failure is also unrelated and also caused non-deterministic concretization. Fix in #45272

**Comment by ax3l** [comment]
Thank you, @haampie - both for the excellent detection (confirming in #45251) and for fixing the WarpX / openPMD-api constraint.

**Pull request merged**